pkg/util/ephemeral/label_matchers.go

```diff
@@ -0,0 +1,223 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ephemeral
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	amlabels "github.com/prometheus/alertmanager/pkg/labels"
+	"github.com/prometheus/prometheus/model/labels"
+	"gopkg.in/yaml.v3"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+// LabelMatchers configures matchers based on which series get marked as ephemeral.
+type LabelMatchers struct {
+	raw      map[Source][]string
+	bySource map[Source]MatcherSetsForSource
+	string   string
+}
+
+type MatcherSetsForSource []matcherSet
+
+// matcherSet is like alertmanager's labels.Matchers but for Prometheus' labels.Matcher slice
+type matcherSet []*labels.Matcher
+
+// matches checks whether all the matchers match the given label set.
+func (ms matcherSet) matches(lset []mimirpb.LabelAdapter) bool {
+	for _, m := range ms {
+		var lv string
+		for _, l := range lset {
+			if l.Name == m.Name {
+				lv = l.Value
+				break
+			}
+		}
+
+		if !m.Matches(lv) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// HasMatchers returns true if there is at least one matcher defined, otherwise it returns false.
+func (m *MatcherSetsForSource) HasMatchers() bool {
+	return len(*m) > 0
+}
+
+func (m *MatcherSetsForSource) ShouldMarkEphemeral(lset []mimirpb.LabelAdapter) bool {
+	for _, ms := range *m {
+		if ms.matches(lset) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// String is a canonical representation of the config, it is compatible with the flag definition.
+// String is needed to implement flag.Value.
+func (c *LabelMatchers) String() string {
+	return c.string
+}
+
+// Set implements flag.Value, and is used to set the config value from a flag value provided as string.
+// Set is needed to implement flag.Value.
+func (c *LabelMatchers) Set(s string) error {
+	if strings.TrimSpace(s) == "" {
```

Comment:
Don't you want to run the rest of the function code in the trimmed version of `s`?

Reply:
actually that wasn't my intention, but now that you ask I think I should, I'll fix it.

Reply:
e55720991